proxy/grpcproxy/cache/store.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2016 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"sync"
+	"sort"
+
+	"github.com/coreos/etcd/client"
+	"github.com/golang/groupcache/lru"
+)
+
+const maxEntries = 2048
+
+type Cache interface {
+	Add(event *client.Response)
+	Get(key string, revision uint64) (*client.Response, error)
+	Compact(revision uint64) error
+}
+
+// DeltaCompactor is an algorithm that removes redundant changes.
+type DeltaCompactor interface {
+	Compact(Deltas) Deltas
+}
+
+// DeltaCompactorFunc should remove redundant changes; but changes that
+// are redundant depend on one's desired semantics, so this is an
+// injectable function.
+//
+// DeltaCompactorFunc adapts a raw function to be a DeltaCompactor.
+type DeltaCompactorFunc func(deltas Deltas, revision uint64) Deltas
+
+// Compress just calls dc.
+func (dc DeltaCompactorFunc) Compress(deltas Deltas, revision uint64) Deltas {
+	return dc(deltas)
+}
+
+func defaultCompactor(deltas Deltas, revision uint64) Deltas {
+	index := sort.Search(len(deltas), func(i int) { return deltas[i] > revision })
+	ret := Deltas{}
+	ret = append(ret, deltas[index:])
+	return ret
+}
+
+// Deltas is a list of one or more 'Delta's to an individual object.
+// The oldest delta is at index 0, the newest delta is the last one.
+type Deltas []Delta
```

Comment:
I think this is a little bit over kill. We already have lru there. We can just let lru take care of everything at day 1, and make it more sophisticated later when we know more about the use case and expectation from users.

So my suggestion is:

1. use key + revision as the cache key in lru.
2. add another map that maps from a key to its most recent revision in lru.

When users get key with a specific revision, we can query the lru cache directly.

When users get key without a specific revision, we find the revision from the map and then query lru.